# Deep Learning - Question 1

### Use the use case in the class https://umkc.box.com/s/3cvfiwc81lhgygc67deyeqs8m858lld0In  :  
1. Add more Dense layers to the existing code and check howthe accuracy changes.

In [1]:
#Importing Libraries
import numpy as np
import pandas as pd
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from sklearn.model_selection import train_test_split

In [2]:
#Importing the dataset
diabetes = pd.read_csv('diabetes.csv',header = None).values

In [3]:
#splitting the data into training and testing categories
X_train, X_test, Y_train, Y_test = train_test_split(diabetes[:,0:8], diabetes[:,8],test_size=0.25, random_state=87)

In [4]:
#Building the sequential model
np.random.seed(155)
my_first_nn = Sequential() # create model
my_first_nn.add(Dense(20, input_dim=8, activation='relu')) # hidden layer
my_first_nn.add(Dense(1, activation='sigmoid')) # output layer

#Compilation 
my_first_nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['acc'])
my_first_nn_fitted = my_first_nn.fit(X_train, Y_train, epochs=100,initial_epoch=0)

#Printing the summary and accuarcy of the model
print(my_first_nn.summary())
print(my_first_nn.evaluate(X_test, Y_test))

Epoch 1/100
18/18 [==============================] - 0s 2ms/step - loss: 32.6237 - acc: 0.3385
Epoch 2/100
18/18 [==============================] - 0s 2ms/step - loss: 18.9617 - acc: 0.3264
Epoch 3/100
18/18 [==============================] - 0s 2ms/step - loss: 8.9081 - acc: 0.3889
Epoch 4/100
18/18 [==============================] - 0s 2ms/step - loss: 5.2283 - acc: 0.5538
Epoch 5/100
18/18 [==============================] - 0s 2ms/step - loss: 3.6317 - acc: 0.5712
Epoch 6/100
18/18 [==============================] - 0s 2ms/step - loss: 3.0671 - acc: 0.5642
Epoch 7/100
18/18 [==============================] - 0s 2ms/step - loss: 2.8352 - acc: 0.5521
Epoch 8/100
18/18 [==============================] - 0s 2ms/step - loss: 2.6094 - acc: 0.5694
Epoch 9/100
18/18 [==============================] - 0s 2ms/step - loss: 2.3725 - acc: 0.5712
Epoch 10/100
18/18 [==============================] - 0s 2ms/step - loss: 2.1177 - acc: 0.5816
Epoch 11/100
18/18 [==============================] - 0s 

18/18 [==============================] - 0s 2ms/step - loss: 0.6106 - acc: 0.7153
Epoch 88/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5742 - acc: 0.7205
Epoch 89/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5634 - acc: 0.7222
Epoch 90/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5780 - acc: 0.7170
Epoch 91/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5673 - acc: 0.7205
Epoch 92/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5802 - acc: 0.7326
Epoch 93/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5806 - acc: 0.7292
Epoch 94/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5581 - acc: 0.7326
Epoch 95/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5687 - acc: 0.7240
Epoch 96/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5707 - acc: 0.7153
Epoch 97/100
18/18 [==============================] - 0s 2ms/st

#### Observations :
1. Loss     = 73%
2. Accuracy = 66%

### Adding more dense layers 

In [5]:
#Building the sequential model
np.random.seed(155)
my_first_nn = Sequential() # create model
my_first_nn.add(Dense(20, input_dim=8, activation='relu')) # hidden layer
my_first_nn.add(Dense(40, activation='relu')) #adding more hidden layers
my_first_nn.add(Dense(1, activation='sigmoid')) # output layer
my_first_nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['acc'])
my_first_nn_fitted = my_first_nn.fit(X_train, Y_train, epochs=100,initial_epoch=0)

print(my_first_nn.summary())
print(my_first_nn.evaluate(X_test, Y_test))

Epoch 1/100
18/18 [==============================] - 0s 2ms/step - loss: 4.5061 - acc: 0.5451
Epoch 2/100
18/18 [==============================] - 0s 2ms/step - loss: 1.5055 - acc: 0.6024
Epoch 3/100
18/18 [==============================] - 0s 2ms/step - loss: 1.1373 - acc: 0.5955
Epoch 4/100
18/18 [==============================] - 0s 2ms/step - loss: 0.9746 - acc: 0.5938
Epoch 5/100
18/18 [==============================] - 0s 2ms/step - loss: 0.8742 - acc: 0.6319
Epoch 6/100
18/18 [==============================] - 0s 2ms/step - loss: 0.8333 - acc: 0.6198
Epoch 7/100
18/18 [==============================] - 0s 2ms/step - loss: 0.7595 - acc: 0.6458
Epoch 8/100
18/18 [==============================] - 0s 2ms/step - loss: 0.7877 - acc: 0.6458
Epoch 9/100
18/18 [==============================] - 0s 2ms/step - loss: 0.7523 - acc: 0.6319
Epoch 10/100
18/18 [==============================] - 0s 2ms/step - loss: 0.7523 - acc: 0.6285
Epoch 11/100
18/18 [==============================] - 0s 2m

18/18 [==============================] - 0s 2ms/step - loss: 0.4749 - acc: 0.7656
Epoch 88/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4761 - acc: 0.7778
Epoch 89/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4758 - acc: 0.7812
Epoch 90/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4846 - acc: 0.7812
Epoch 91/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4699 - acc: 0.7865
Epoch 92/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4939 - acc: 0.7587
Epoch 93/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4894 - acc: 0.7500
Epoch 94/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4835 - acc: 0.7535
Epoch 95/100
18/18 [==============================] - 0s 2ms/step - loss: 0.5066 - acc: 0.7604
Epoch 96/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4854 - acc: 0.7691
Epoch 97/100
18/18 [==============================] - 0s 2ms/st

#### Observations :
1. Loss     = 66%
2. Accuracy = 69%

### Adding more dense layers

In [6]:
np.random.seed(155)
my_first_nn = Sequential() # create model
my_first_nn.add(Dense(20, input_dim=8, activation='relu')) # hidden layer
my_first_nn.add(Dense(40, activation='relu')) #adding more hidden layers
my_first_nn.add(Dense(60, activation='relu')) #adding more hidden layers
my_first_nn.add(Dense(1, activation='sigmoid')) # output layer
my_first_nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['acc'])
my_first_nn_fitted = my_first_nn.fit(X_train, Y_train, epochs=100,initial_epoch=0)

print(my_first_nn.summary())
print(my_first_nn.evaluate(X_test, Y_test))

Epoch 1/100
18/18 [==============================] - 0s 2ms/step - loss: 2.6469 - acc: 0.6024
Epoch 2/100
18/18 [==============================] - 0s 2ms/step - loss: 1.0909 - acc: 0.5972
Epoch 3/100
18/18 [==============================] - 0s 2ms/step - loss: 0.8291 - acc: 0.6510
Epoch 4/100
18/18 [==============================] - 0s 2ms/step - loss: 0.7216 - acc: 0.6562
Epoch 5/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6441 - acc: 0.6736
Epoch 6/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6412 - acc: 0.6597
Epoch 7/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6339 - acc: 0.6753
Epoch 8/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6236 - acc: 0.6910
Epoch 9/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6268 - acc: 0.6806
Epoch 10/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6592 - acc: 0.6806
Epoch 11/100
18/18 [==============================] - 0s 2m

18/18 [==============================] - 0s 2ms/step - loss: 0.4510 - acc: 0.7934
Epoch 88/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4555 - acc: 0.7865
Epoch 89/100
18/18 [==============================] - 0s 3ms/step - loss: 0.4759 - acc: 0.7795
Epoch 90/100
18/18 [==============================] - 0s 3ms/step - loss: 0.5269 - acc: 0.7413
Epoch 91/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4831 - acc: 0.7830
Epoch 92/100
18/18 [==============================] - 0s 3ms/step - loss: 0.4750 - acc: 0.7865
Epoch 93/100
18/18 [==============================] - 0s 3ms/step - loss: 0.4523 - acc: 0.7917
Epoch 94/100
18/18 [==============================] - 0s 2ms/step - loss: 0.4466 - acc: 0.7830
Epoch 95/100
18/18 [==============================] - 0s 3ms/step - loss: 0.4436 - acc: 0.7917
Epoch 96/100
18/18 [==============================] - 0s 3ms/step - loss: 0.4600 - acc: 0.7986
Epoch 97/100
18/18 [==============================] - 0s 2ms/st

#### Observations :
1. Loss     = 61%
2. Accuracy = 72%

### Adding more dense layers

In [7]:
np.random.seed(155)
my_first_nn = Sequential() # creating the model
my_first_nn.add(Dense(20, input_dim=8, activation='relu')) # hidden layer
my_first_nn.add(Dense(40, activation='relu')) #adding more hidden layers
my_first_nn.add(Dense(60, activation='relu')) #adding more hidden layers
my_first_nn.add(Dense(100, activation='relu')) #adding more hidden layers
my_first_nn.add(Dense(1, activation='sigmoid')) # output layer
my_first_nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['acc'])
my_first_nn_fitted = my_first_nn.fit(X_train, Y_train, epochs=100,initial_epoch=0)

print(my_first_nn.summary())
print(my_first_nn.evaluate(X_test, Y_test))

Epoch 1/100
18/18 [==============================] - 0s 3ms/step - loss: 1.2550 - acc: 0.5556
Epoch 2/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6901 - acc: 0.6649
Epoch 3/100
18/18 [==============================] - 0s 3ms/step - loss: 0.6229 - acc: 0.6840
Epoch 4/100
18/18 [==============================] - 0s 3ms/step - loss: 0.6147 - acc: 0.6736
Epoch 5/100
18/18 [==============================] - 0s 3ms/step - loss: 0.5936 - acc: 0.6858
Epoch 6/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6063 - acc: 0.6997
Epoch 7/100
18/18 [==============================] - 0s 3ms/step - loss: 0.6051 - acc: 0.6892
Epoch 8/100
18/18 [==============================] - 0s 2ms/step - loss: 0.6006 - acc: 0.6944
Epoch 9/100
18/18 [==============================] - 0s 3ms/step - loss: 0.5867 - acc: 0.7031
Epoch 10/100
18/18 [==============================] - 0s 3ms/step - loss: 0.5619 - acc: 0.7170
Epoch 11/100
18/18 [==============================] - 0s 2m

18/18 [==============================] - 0s 3ms/step - loss: 0.3694 - acc: 0.8160
Epoch 88/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3517 - acc: 0.8438
Epoch 89/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3606 - acc: 0.8299
Epoch 90/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3508 - acc: 0.8403
Epoch 91/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3632 - acc: 0.8438
Epoch 92/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3257 - acc: 0.8524
Epoch 93/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3324 - acc: 0.8385
Epoch 94/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3398 - acc: 0.8368
Epoch 95/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3543 - acc: 0.8403
Epoch 96/100
18/18 [==============================] - 0s 3ms/step - loss: 0.3292 - acc: 0.8576
Epoch 97/100
18/18 [==============================] - 0s 3ms/st

#### Observations :
1. Loss     = 74%
2. Accuracy = 73%

## Conclussion :
1. Accuracy with one hidden layer   : 66 %
2. Accuracy with two hidden layer   : 69 %
3. Accuracy with three hidden layer : 72 %
4. Accuracy with four hidden layer  : 73 %

<h3> Accuracy is increasing with more dense hidden layers </h3>